http/server.ts

```diff
@@ -1,111 +1,117 @@
 // Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
-import { listen, Conn, toAsyncIterator, Reader, Writer, copy } from "deno";
-import { BufReader, BufState, BufWriter } from "../io/bufio.ts";
+
+import { Conn, copy, listen, Reader, toAsyncIterator, Writer } from "deno";
+import { BufReader, BufWriter } from "../io/bufio.ts";
 import { TextProtoReader } from "../textproto/mod.ts";
 import { STATUS_TEXT } from "./http_status.ts";
 import { assert } from "../testing/mod.ts";
+import { defer, Deferred } from "../async/deferred.ts";
+import { Key, pathToRegexp } from "./path_to_regexp.ts";
+import { BodyReader, ChunkedBodyReader } from "./readers.ts";
 
-interface Deferred {
-  promise: Promise<{}>;
-  resolve: () => void;
-  reject: () => void;
-}
+export type HttpHandler = (req: ServerRequest) => Promise<any>;
 
-function deferred(): Deferred {
-  let resolve, reject;
-  const promise = new Promise((res, rej) => {
-    resolve = res;
-    reject = rej;
-  });
-  return {
-    promise,
-    resolve,
-    reject
-  };
-}
-
-interface ServeEnv {
-  reqQueue: ServerRequest[];
-  serveDeferred: Deferred;
-}
+export type ServerRequest = {
+  url: string;
+  method: string;
+  proto: string;
+  headers: Headers;
+  params: { [key: string]: string };
+  body: Reader;
+  respond: ServerResponder;
+};
 
-/** Continuously read more requests from conn until EOF
- * Calls maybeHandleReq.
- * bufr is empty on a fresh TCP connection.
- * Would be passed around and reused for later request on same conn
- * TODO: make them async function after this change is done
- * https://github.com/tc39/ecma262/pull/1250
- * See https://v8.dev/blog/fast-async
- */
-function serveConn(env: ServeEnv, conn: Conn, bufr?: BufReader) {
-  readRequest(conn, bufr).then(maybeHandleReq.bind(null, env, conn));
-}
+export type ServerResponder = (response: ServerResponse) => Promise<any>;
 
-function maybeHandleReq(env: ServeEnv, conn: Conn, maybeReq: any) {
-  const [req, _err] = maybeReq;
-  if (_err) {
-    conn.close(); // assume EOF for now...
-    return;
-  }
-  env.reqQueue.push(req); // push req to queue
-  env.serveDeferred.resolve(); // signal while loop to process it
+export interface ServerResponse {
+  status?: number;
+  headers?: Headers;
+  body?: Uint8Array | Reader;
 }
 
-export async function* serve(addr: string) {
+export async function* serve(
```

Comment:
After some more investigation I'm skeptical about rewriting this function (minus `readRequest`/`writeResponse` bit) right now. We still need to figure out how to reuse connections and support keep-alive and this solution doesn't solve these problems. We should discuss and cross check in other languages.

Reply:
I reverted implementation of `serve()` as possible. But race for cancellation may inevitably make performance degraded.